src/nvim/eval.c

```diff
@@ -8495,6 +8495,54 @@ static void f_empty(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   rettv->vval.v_number = n;
 }
 
+/// "environ()" function
+static void f_environ(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  int i = 0;
+  char_u *entry, *value;
+# ifdef WIN32
+  extern wchar_t **_wenviron;
+# else
+  extern char **environ;
+# endif
+
+  tv_dict_alloc_ret(rettv);
+
+# ifdef WIN32
+  if (*_wenviron == NULL) {
+    return;
+  }
+# else
+  if (*environ == NULL) {
+    return;
+  }
+# endif
+
+  for (i = 0; ; i++) {
+# ifdef WIN32
+    uint16_t *p;
+
+    if ((p = (uint16_t *)_wenviron[i]) == NULL) {
+      return;
+    }
+    entry = utf16_to_enc(p, NULL);
+# else
+    if ((entry = (char_u *)environ[i]) == NULL) {
+      return;
+    }
+    entry = vim_strsave(entry);
+# endif
```

Comment:
maybe `os_getenvname_at_index` can be used instead of `_wenviron` / `environ` ? It's inefficient, but saves code and avoids bugs.

Reply:
Yeah, came across it just now also - but wanted to state that it _is_ really inefficient.
Will give it a try.

Reply:
.. `os_getenv` returning NULL for empty is a problem here also - known since 2015: https://github.com/neovim/neovim/issues/3266